lib/Hakyll/Core/Compiler/Internal.hs

```diff
@@ -112,18 +114,19 @@ instance Monoid CompilerWrite where
 
 --------------------------------------------------------------------------------
 -- | Distinguishes reasons in a 'CompilerError'
-data Reason a
-    -- | An exception occured during compilation
-    = CompilationFailure a
-    -- | Absence of any result, most notably in template contexts
-    | NoCompilationResult a
+data CompilerErrors a
+    -- | One or more exceptions occured during compilation
+    = CompilationFailure (NonEmpty a)
```

Comment:
What's your specific reasoning to avoid empty lists here? And why did `compilerThrow` change to produce a `CompilationNoResult` if no reason value is given?
(Not assuming it's a bad thing, I just don't understand the idea behind it).

Reply:
Also seems to make the `Alternative.<|>` implementation more complicated

Reply:
It's to prevent us from shooting ourselves in the foot -- if there's no error, we shouldn't get into a `CompilationFailure` scenario.  This is not true for the other case where we have `empty`.

Reply:
After familiarising myself with the code again, I would heavily prefer
```hs
-- | Put a 'CompilerError' with  multiple error messages as 'CompilationFailure'
compilerThrow :: NonEmpty String -> Compiler a
compilerThrow = compilerResult . CompilerError . CompilationFailure
```
The `compilerThrow` function is an internal function (only available on `Compiler.Internal`) so we can easily "break" it by changing the type. (IIRC it was earlier used in `Context` or so for transforming errors, but has been replaced there with more sensible alternatives). It never was meant to be called with an empty list, and neither in `fail` nor in `<|>` that could happen.
By having `compilerThrow :: NonEmpty String -> Compiler a`, we don't need to convert back and forth between lists and non-empty lists in `<|>`.

Where it does matter, and where I agree with you that we should prevent anyone from shooting themselves in the foot, is `throwError` in the `MonadError [String]` instance. Let's do
```
throwError errs = maybe (compilerNoResult errs) compilerThrow $ NonEmpty.nonEmpty errs
```